### PR TITLE
Shard on read Dtensor aware

### PR DESCRIPTION
**Pull request by 3outeille**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46717)

# What does this PR do ?

This PR adds `shard-on-read` support for loading DTensor-sharded model parameters during `from_pretrained`. Instead of materializing the full checkpoint tensor on every rank and sharding afterward, each rank reads only the slice it needs from the checkpoint, based on the target parameter’s DTensor mesh and placements.

For now I don't want to break the integration with the existing TP on main (since we haven't merged the inference-compatible DTensor TP yet). So here's what I'm doing:

- I'm doing 2 loading paths for the moment (Dtensor and the old one for TP only)
- I fully integrate FSDP and get it working with shard-on-read to keep the distributed PR as small as possible
- Once only TP is left, I remove the second TP loading path and add DTensor inference for TP

### CODE REVIEW

File: src/transformers/distributed/sharding_utils.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
+# Copyright 2025 The HuggingFace Team. All rights reserved.
```
Comment (3outeille) [inline src/transformers/distributed/sharding_utils.py]: 2026

File: src/transformers/distributed/sharding_utils.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                if not placement.is_shard():
+                    intervals[source_dim] = self._strided_intervals(
+                        intervals[source_dim], rank, world_size, placement.split_factor
+                    )
+                else:
+                    intervals[source_dim] = self._contiguous_intervals(intervals[source_dim], rank, world_size)
+            # Only _StridedShard can produce multi-interval dims that need cat.
+            if has_strided:
+                return self._slice_and_cat(source, intervals, device, dtype)
+            slices = tuple(slice(*(pieces[0] if pieces else (0, 0))) for pieces in intervals)
```
Comment (3outeille) [inline src/transformers/distributed/sharding_utils.py]: double checking this part

File: src/transformers/distributed/sharding_utils.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        # form this rank's local shard of the param.
+        shards_leading_axis = any(self._norm_dim(p.dim) == 0 for _, p in placements)
+        owns_index = self._axis0_offset <= tensor_idx < self._axis0_offset + self._axis0_local_size
+        if shards_leading_axis and not owns_index:
+            return None
+
+        # Inner dims use only _contiguous_intervals (one piece per dim), so a
+        # single slice suffices
+        inner_placements = [(md, p) for md, p in placements if self._norm_dim(p.dim) != 0]
+        if not inner_placements:
```
Comment (3outeille) [inline src/transformers/distributed/sharding_utils.py]: maybe just `not placements` instead ?

File: src/transformers/core_model_loading.py [hunk 4]
```diff
@@ -1112,22 +1109,28 @@ def set_param_for_module(
...
             expected_shape = torch.Size(distributed_operation.get_expected_sharded_shape(ref.shape))
+        elif isinstance(ref, DTensor):
+            expected_shape = ref._local_tensor.shape
         else:
             expected_shape = ref.shape
 
         if ref is not None and param_value.shape != expected_shape and hf_quantizer is None:
             loading_info.mismatched_keys.add((target_name, param_value.shape, expected_shape))
         else:
+            if isinstance(ref, DTensor):
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py:1324]:
```suggestion
elif isinstance(ref, DTensor):
```

File: src/transformers/core_model_loading.py [hunk 5]
```diff
@@ -1439,17 +1458,16 @@ def convert_and_load_state_dict_in_model(
                         and any(isinstance(op, MergeModulelist) for op in mapping.operations)
                         else None
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]: let's compute this only once

File: src/transformers/core_model_loading.py [hunk 6]
```diff
@@ -1439,17 +1458,16 @@ def convert_and_load_state_dict_in_model(
...
                     )
-                    future_or_tensor = spawn_tp_materialize(
+                    future_or_tensor = spawn_materialize(
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]: this call can probably be done only once no? with correct arguments

File: tests/utils/test_core_model_loading.py [hunk 7]
```diff
@@ -225,6 +228,99 @@ def __init__(self, config, add_extra_moe=False, with_mlp=True):
...
+        The model stores them packed into one tensor::
+
+            experts.gate_up_proj.weight  (2, 4, 2)
+                                          ^  ^  ^
+                                          |  |  +-- features
+                                          |  +-- w1 (2) + w3 (2) concatenated
+                                          +-- num_experts
+
+        The conversion (without FSDP) is: load all expert w1/w3 tensors,
+        MergeModulelist(dim=0) stacks experts, Concatenate(dim=1) joins w1+w3.
```
Comment (ArthurZucker) [inline tests/utils/test_core_model_loading.py]: This depends, on TP or EP. 
EP loads only num_experts // ranks 
TP load num_experts, w1 // ranks

File: tests/utils/test_core_model_loading.py [hunk 8]
```diff
@@ -225,6 +228,99 @@ def __init__(self, config, add_extra_moe=False, with_mlp=True):
...
+
+        With FSDP, Shard(0) splits the expert dim across ranks. Rank 0 owns
+        expert 0, rank 1 owns expert 1. So rank 0 should skip loading expert 1
+        entirely -- not load it then discard it.
```
Comment (ArthurZucker) [inline tests/utils/test_core_model_loading.py]: this never happened before !

File: tests/utils/test_core_model_loading.py [hunk 9]
```diff
@@ -225,6 +228,99 @@ def __init__(self, config, add_extra_moe=False, with_mlp=True):
 
 
 class TestConvertAndLoadStateDict(unittest.TestCase):
+    def test_dtensor_shard_aware_mixtral_conversion_uses_only_local_experts(self):
```
Comment (ArthurZucker) [inline tests/utils/test_core_model_loading.py:246]: I like this small and minimal test!

File: src/transformers/distributed/sharding_utils.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+    from torch.distributed.tensor.placement_types import Shard
+
+    # torch < 2.10 names as an underscore before `local_shard_size_and_offset`: alias it the non-underscored version
+    if not hasattr(Shard, "local_shard_size_and_offset") and hasattr(Shard, "_local_shard_size_and_offset"):
+        Shard.local_shard_size_and_offset = Shard._local_shard_size_and_offset
+
+
+class DtensorShardOperation:
+    """Shard-on-read: slice a full checkpoint tensor down to this rank's local
+    DTensor shard, for any combination of placements on a 1-D or n-D mesh.
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]:
```suggestion
    """Shard-on-read: slice a full disk tensor down to this rank's local
    DTensor shard, for any combination of placements on a 1-D or n-D mesh. It's on read because instructions are made so the cpu only fetches on disk the parts we want.
```

File: src/transformers/distributed/sharding_utils.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+
+    Different scenarios of different placements
+    ------------------------------------------
+    Placement tuples are ordered outermost-first; for a 2-D (fsdp, tp) mesh
+    the tuple is (fsdp_placement, tp_placement).
+
+    | Scenario                                          | Placements                                  |
+    |---------------------------------------------------|---------------------------------------------|
+    | TP-only, non-fused (e.g. q_proj/k_proj/v_proj)    | [Shard(d)]                                   |
+    | TP-only, fused gate/up                            | [_StridedShard(d, sf=2)]                     |
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py:61]: this is only gpt OSS, the normal fused gate up is just Shard(1) !

File: src/transformers/distributed/sharding_utils.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+    | TP + FSDP, same tensor dim (contiguous TP case)   | [Shard(d), Shard(d)]                         |
+    | TP + FSDP, same tensor dim (fused/interleaved TP) | [_StridedShard(d, sf=tp_size), Shard(d)]     |
+    | TP + FSDP, different dims                         | [Shard(d1), Shard(d2)]                       |
+
+    Mesh dimensions are listed outermost-first. For a 2-D (fsdp=F, tp=T) mesh,
+    rank index = fsdp_idx * T + tp_idx.
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]: I don't understand what T is , nor F, or why a rank index deeds fsdp / tp index

File: src/transformers/distributed/sharding_utils.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+    Loading (this class)
+    --------------------
+    During `from_pretrained`, each rank reads the full checkpoint tensors,
+    then calls `shard_tensor(source, tensor_idx=...)` to keep only its local
+    DTensor shard. The class encapsulates the placements + mesh so the
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]: No, this is not true! 
Otherwise this is `read_then_shard` mechanism

File: src/transformers/distributed/sharding_utils.py [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+    slicing logic doesn't have to be repeated at every call site.
+
+    Two checkpoint layouts are supported. Running example: an MoE weight
+    stack of param shape [N_experts, in, out]:
+
+    | Checkpoint layout                  | tensor_idx | source.shape  | Returns                     |
+    |------------------------------------|------------|---------------|-----------------------------|
+    | One stacked tensor                 | None       | [N, in, out]  | This rank's slice along     |
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]: this is just that we want to shard on in or out dim

File: src/transformers/distributed/sharding_utils.py [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+
+    Loading (this class)
+    --------------------
+    During `from_pretrained`, each rank reads the full checkpoint tensors,
+    then calls `shard_tensor(source, tensor_idx=...)` to keep only its local
+    DTensor shard. The class encapsulates the placements + mesh so the
+    slicing logic doesn't have to be repeated at every call site.
+
+    Two checkpoint layouts are supported. Running example: an MoE weight
+    stack of param shape [N_experts, in, out]:
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]: mention why, they com from the weight loader

File: src/transformers/distributed/sharding_utils.py [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+            # Determine for each tensor dimension, which type of sharding operations to apply (_StridedShard or Shard) and which rank to apply it to.
+            # i.e: dim 0 -> [ Strided(rank0, size=2, sf=2), Shard(rank1, size=2) ]
+            # i.e: dim 1 -> [ Shard(rank0, size=2)]
+            planned_ops_by_dim = [[] for _ in source_shape]
+            for mesh_dim, placement in dim_placements:
+                sub_mesh = self._get_sub_mesh(mesh_dim)
+                rank, world_size = sub_mesh.get_local_rank(), sub_mesh.size()
+                dim_idx = self._normalize_param_dim(placement.dim)
+                planned_ops_by_dim[dim_idx].append((placement, rank, world_size))
+
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py:125]: we should probably not do this on the hot path no?

File: src/transformers/distributed/sharding_utils.py [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+            # Apply the sharding operations to each tensor dimension.
+            intervals_by_dim = [[(0, size)] for size in source_shape]
+            for dim_idx, planned_ops in enumerate(planned_ops_by_dim):
+                intervals = intervals_by_dim[dim_idx]
+                for placement, rank, world_size in planned_ops:
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py:130]: we need a test where a dim is shared twice!

File: src/transformers/distributed/sharding_utils.py [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+            # i.e: dim 0 -> [ Strided(rank0, size=2, sf=2), Shard(rank1, size=2) ]
+            # i.e: dim 1 -> [ Shard(rank0, size=2)]
+            planned_ops_by_dim = [[] for _ in source_shape]
+            for mesh_dim, placement in dim_placements:
+                sub_mesh = self._get_sub_mesh(mesh_dim)
+                rank, world_size = sub_mesh.get_local_rank(), sub_mesh.size()
+                dim_idx = self._normalize_param_dim(placement.dim)
+                planned_ops_by_dim[dim_idx].append((placement, rank, world_size))
+
+            # Apply the sharding operations to each tensor dimension.
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]:
```suggestion
            # prepare the slices to fetch on disk
```

File: src/transformers/distributed/sharding_utils.py [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+                intervals = intervals_by_dim[dim_idx]
+                for placement, rank, world_size in planned_ops:
+                    if placement.is_shard():
+                        intervals = self._apply_contiguous_shard(intervals, rank, world_size)
+                    else:
+                        intervals = self._apply_strided_shard(intervals, rank, world_size, placement.split_factor)
+                intervals_by_dim[dim_idx] = intervals
+
+            has_strided_shard = any(not placement.is_shard() for _, placement in dim_placements)
+            if has_strided_shard:
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py:140]:
```suggestion
            # finally fetch from the disk only the slices
            if has_strided_shard:
```

File: src/transformers/distributed/sharding_utils.py [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+                # Multi-interval dim: read each piece separately, then concatenate.
+                return self._slice_and_cat(source, intervals_by_dim, device, dtype)
+            else:
+                slice_parts = []
+                for intervals in intervals_by_dim:
+                    start, end = intervals[0] if len(intervals) > 0 else (0, 0)
+                    slice_parts.append(slice(start, end))
+
+                return source[tuple(slice_parts)].to(device=device, dtype=dtype)
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py:149]: we need to have some kind of ID / Dtensor to potentially preven re-computing this for all layers for example! might be a todo

File: src/transformers/distributed/sharding_utils.py [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+                source_dim = param_dim - 1
+                sub_mesh = self._get_sub_mesh(mesh_dim)
+                rank, world_size = sub_mesh.get_local_rank(), sub_mesh.size()
+                planned_ops_by_source_dim[source_dim].append((rank, world_size))
+
+        intervals_by_source_dim = [[(0, size)] for size in source_shape]
+        for source_dim, planned_ops in enumerate(planned_ops_by_source_dim):
+            intervals = intervals_by_source_dim[source_dim]
+            for rank, world_size in planned_ops:
+                intervals = self._apply_contiguous_shard(intervals, rank, world_size)
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]:
```suggestion
                intervals = self._compute_contiguous_slice(intervals, rank, world_size)
```

**Review by ArthurZucker — COMMENTED** [review]
cool!

File: src/transformers/core_model_loading.py (hunk 6 above)
Comment (3outeille) [reply src/transformers/core_model_loading.py]: this can be called once but I wanted to make the distinction clear to introduce minimal changes to the PR

File: src/transformers/core_model_loading.py (hunk 4 above)
Comment (3outeille) [reply src/transformers/core_model_loading.py:1324]: if we do that, we will have to do 

```
mismatch = ref is not None and param_value.shape != expected_shape and hf_quantizer is None

if mismatch:
    loading_info.mismatched_keys.add(...)
else:
    if isinstance(ref, DTensor):
        ...

if not mismatch:
    param_value._is_hf_initialized = True
    setattr(...)
```

which is the same

File: src/transformers/distributed/sharding_utils.py (hunk 11 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py:61]: it's the other way around, Mixtral like models have `_StridedShard` (that's why we have `get_packed_weights` to make sure each rank has their respective interleaved chunk)

On the other hand, gpt-oss is the only one to have `Shard` because in the modeling we have `gate, up = gate_up[..., ::2], gate_up[..., 1::2]` that handles the interleaving

File: src/transformers/distributed/sharding_utils.py (hunk 16 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py:125]: provided that you have at top 3 dimensions (`[N, in, out]`), that is very negligeable as you will loop just 3 times max

File: src/transformers/distributed/sharding_utils.py (hunk 12 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py]: will remove this sentence, it doesn't bring any useful information to the user

File: src/transformers/distributed/sharding_utils.py (hunk 13 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py]: rephrasing

File: src/transformers/distributed/sharding_utils.py (hunk 14 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py]: I added a new doc comment (as asked above) which made this table redundant

File: src/transformers/distributed/sharding_utils.py (hunk 17 above)
Comment (3outeille) [reply src/transformers/distributed/sharding_utils.py:130]: We have it already at `test_2D_shard_same_dim` in `tests/utils/test_distributed_sharding_utils.py`

File: tests/utils/test_core_model_loading.py (hunk 7 above)
Comment (3outeille) [reply tests/utils/test_core_model_loading.py]: added more details to it

File: tests/utils/test_core_model_loading.py (hunk 8 above)
Comment (3outeille) [reply tests/utils/test_core_model_loading.py]: wrong phrasing, I wanted to say something about FSDP sharding the same dimension than TP/EP and how it is handled in 2D-parallelism (which wasn't handle before). Will rephrase

File: src/transformers/core_model_loading.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/core_model_loading.py:1324]: okok

File: src/transformers/core_model_loading.py [hunk 22]
```diff
@@ -1448,27 +1642,17 @@ def convert_and_load_state_dict_in_model(
...
-                        sharding_op=mapping.distributed_operation,
-                        tensor_idx=shard_index,
-                    )
-
-            if future_or_tensor is None:
-                future_or_tensor = spawn_materialize(thread_pool, tensor, param_device, _dtype)
+                    sharding_op = mapping.distributed_operation
+                    materialize_device = device_map[""]
+
+            future_or_tensor = spawn_materialize(
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py:1648]: ty

**Review by ArthurZucker — APPROVED** [review]
ty
